Improve C# intro docs/Unity transition guide

mhilbrunner opened the pull request:
Some small additions to the new C# docs.

Also linked them in the Unity transition docs and cleaned that up a little, fixed the links to physics and some small corrections. Removed outdated references to UnityScript/JavaScript. Also added links to VisualScripting and GDScript docs.

The PNG was optimized with a lossy encoder.

Also re-ordered the C# intro to make more sense:
1. Intro to C#
2. Setting up C#
3. Example
4. History at end instead of between 2 and 3

skyace65 commented:
I personally think the History section was fine where it was but that's just me. The rest of the updates to the C# page look great, it's much more detailed than it originally was and should be extremely useful to C# and Godot newcomers. 

I think it might be worth renaming the scripting section in the Unity to Godot docs to something like "Scripting: GDScript, C# and Visual Script. And it looks like there's a conflict with that page that needs to be resolved

mhilbrunner commented:
History should be at the beginning then, instead of between the setup/code stuff.

But IMO, that whole section really belongs in a Credits/Trivia section or something, not in the Intro-to-C#. I even thought about whether it maybe should be dropped entirely. If I look at the docs page for a feature, I don't particularly care about who added it when and why and by whom it was financed, but how it works. And only how it works, as concise, precise and easy to grasp as possible.

Thats just my opininion, however, maybe some other opininions on this? My compromise was to drop it to the end where it is easy to ignore when not interested, but maybe someone has a better suggestion?

Will see if I can resolve the conflict later :)

mhilbrunner commented:
Fixed the merge conflict and another typo I found, and changed the Scripting section title in Unity to Godot.

RabbitB commented:
There looks to be a typo in `unity_to_godot.rst`, line 160.

> between one evening for an experimented programmer and a week for a complete beginner.

Looks like "experimented" should be "experienced".

mhilbrunner commented:
Fixed above typo on that page, thanks Rabbit :)

akien-mga commented on learning/editor/unity_to_godot.rst:
```diff
@@ -149,19 +149,19 @@ Add a component -> Inheriting
 
-Scripting : From C# to GDScript
+Scripting: GDScript, C# and Visual Scripting
 -------------------------------
```
You need to keep the underline at the same length as the title.

akien-mga commented on learning/scripting/c_sharp/c_sharp_basics.rst:
```diff
@@ -3,47 +3,89 @@
 C#
 ===
```
Can you fix all underlines in this page? They must be triggering tons of warnings.

akien-mga commented:
> But IMO, that whole section really belongs in a Credits/Trivia section or something, not in the Intro-to-C#. I even thought about whether it maybe should be dropped entirely. If I look at the docs page for a feature, I don't particularly care about who added it when and why and by whom it was financed, but how it works. And only how it works, as concise, precise and easy to grasp as possible.

Agreed, it can be removed IMO.

mhilbrunner commented:
No, that was the only one that triggered any warning - I think its fine when the underlines are too long, which is why they didn't trigger any warnings? Changed all underlines to be exactly as long as their title, regardless. Also dropped the history section.

mhilbrunner commented:
Will update tomorrow, no merge please :)

mhilbrunner commented:
Updated!

Besides updating the Unity transition guide and C# section, the C# scripting section now has the following subpages:
- Introduction
- Features
- API differences to GDScript

Please review, should be fine to merge now if I haven't missed anything :)

neikeq commented on learning/scripting/c_sharp/c_sharp_differences.rst:
```diff
@@ -0,0 +1,213 @@
+Global Scope
+------------
+
+Available under ``Godot.GD`` (formerly ``GodotEngine.Godot``).
```
IMO `(formerly ``GodotEngine.Godot``)` is not necessary.

neikeq commented on learning/scripting/c_sharp/c_sharp_differences.rst:
```diff
@@ -0,0 +1,213 @@
+Available under ``Godot.GD`` (formerly ``GodotEngine.Godot``).
+
+Global functions like ``print``, ``var2str`` and ``weakref`` are located under
+``GD`` in C#.
```
Except math related functions which are in `Godot.Mathf`.

neikeq commented on learning/scripting/c_sharp/c_sharp_differences.rst:
```diff
@@ -0,0 +1,213 @@
+  str.match("*.txt"); // Case sensitive
+  str.expr_match("*.txt", true); // Case sensitive
+  str.matchn("*.txt"); // Case insensitive
+  str.expr_match("*.txt", false); // Case insensitive
```
From `Singletons` to this point, the methods should be in PascalCase.

mhilbrunner commented:
@neikeq Thanks for having a look, fixed those :)

neikeq commented on learning/scripting/c_sharp/c_sharp_features.rst:
```diff
@@ -0,0 +1,72 @@
+.. code-block:: csharp
+
+    Sprite my_sprite = get_node("MySprite") as Sprite;
+    my_sprite.set_frame(0);
```
Here I would explain it this way:

- Normal cast: `(Sprite)GetNode("MySprite")`. Throws `InvalidCastException` if the returned node cannot be casted to Sprite. You would use it instead of the `as` operator if you are pretty sure it won't fail.
- `as` operator: `GetNode("MySprite") as Sprite`. Returns null if the node cannot be casted to Sprite, and for this reason it cannot be used with value types.
- `is` operator: `GetNode("MySprite") is Sprite`. Returns false if the node cannot be casted to Sprite, otherwise true.

Finally, I would mention pattern matching and link to https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching for more details.

mhilbrunner commented:
@neikeq Good call, updated the casting tutorial with your suggestions

neikeq commented on learning/scripting/c_sharp/c_sharp_features.rst:
```diff
@@ -0,0 +1,93 @@
+
+    Sprite mySprite = GetNode("MySprite") as Sprite;
+    // You should check whether mySprite is null here
+    mySprite.SetFrame(0);
```
Hmm, I would just do the null check rather than writing that comment. One-liner: `mySprite?.SetFrame(0);`.

mhilbrunner commented:
@neikeq Updated. Will add the WeakRef/NativeInstance docs in https://github.com/godotengine/godot-docs/issues/580, do you think this is fine now? Or anything else to fix? :)
